contracts/TwabRewards.sol

```diff
@@ -0,0 +1,379 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.6;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@pooltogether/v4-core/contracts/interfaces/ITicket.sol";
+import "@pooltogether/owner-manager-contracts/contracts/Manageable.sol";
+
+import "./interfaces/ITwabRewards.sol";
+
+/**
+ * @title PoolTogether V4 TwabRewards
+ * @author PoolTogether Inc Team
+ * @notice Contract to distribute rewards to depositors in a pool.
+ */
+contract TwabRewards is ITwabRewards, Manageable {
+    using SafeERC20 for IERC20;
+
+    /* ============ Global Variables ============ */
+
+    /// @notice Settings of each promotion.
+    mapping(uint256 => Promotion) internal _promotions;
+
+    /// @notice Current promotion id.
+    uint256 internal _currentPromotionId;
+
+    /// @notice Keeps track of claimed rewards per user.
+    /// @dev _claimedRewards[promotionId][epochId][user] => uint256
+    mapping(uint256 => mapping(uint256 => mapping(address => uint256))) internal _claimedRewards;
```

Comment:
Something I just realized with this, is that we only need to store a boolean per epoch.

Once an epoch has passed it will never change, so you just need to record whether a user has claimed for it.

This means two things:

1. You can use a single word to store 256 epochs.
2. Number of epochs must be less than 256.  Even when users extend them.

This will save a tremendous amount of gas.